<?xml version="1.0" ?>

<robot name="robot" xmlns:xacro="http://ros.org/wiki/xacro">
  <xacro:include filename="$(find rrt_exploration_tutorials)/launch/includes/urdf/common_properties.urdf.xacro"/>
  <xacro:include filename="$(find rrt_exploration_tutorials)/launch/includes/urdf/robot3.gazebo.xacro"/>
 
    <link name="base_footprint"/>

    <joint name="base_joint" type="fixed">
      <origin xyz="0 0 0.075" rpy="0 0 0" />
      <parent link="base_footprint"/>
      <child link="base_link" />
    </joint>

    <link name="base_link">
      <visual>
        <geometry>
	  <box size="0.25 0.15 0.005" />
        </geometry>
        <material name="green">
          <color rgba="0.0 1 0.0 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="0.25 0.15 0.005" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.5" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>

    <!-- SECOND_base -->

    <joint name="base_second_joint" type="fixed">
      <origin xyz="0 0 0.06" rpy="0 0 0" />
      <parent link="base_link"/>
      <child link="base_second_link"/>
    </joint>
    <link name="base_second_link">
      <visual>
        <geometry>
	  <box size="0.25 0.15 0.005" />
	  
        </geometry>
        <material name="green">
          <color rgba="0.0 1 0.0 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="0.25 0.15 0.005" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.5" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>
    

    <!-- wheels -->

    <joint name="front_left_wheel_joint" type="continuous">
      <parent link="front_left_motor_link"/>
      <child link="front_left_wheel_link"/>
      <origin xyz="0.0 0.045 -0.015" rpy="1.57 0 0"/>
      <axis xyz="0 0 1"/>
    </joint>
    <link name="front_left_wheel_link">
      <visual>
        <geometry>
          <mesh filename="package://rrt_exploration_tutorials/launch/includes/meshes/wheel.dae"/>
        </geometry>
        <material name="blue">
          <color rgba="0.0 0.0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder length="0.0206" radius="0.0352"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="0.5" />
        <inertia ixx="0.05" ixy="0" ixz="0" iyy="0.05" iyz="0.0" izz="0.05" />
      </inertial>
    </link>
    
    <joint name="front_right_wheel_joint" type="continuous">
      <parent link="front_right_motor_link"/>
      <child link="front_right_wheel_link"/>
      <origin xyz="0.0 -0.045 -0.015" rpy="1.57 0 0"/>
      <axis xyz="0 0 1"/>
    </joint>
    <link name="front_right_wheel_link">
      <visual>
        <geometry>
          <mesh filename="package://rrt_exploration_tutorials/launch/includes/meshes/wheel.dae"/>
        </geometry>
        <material name="blue">
          <color rgba="0.0 0.0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder length="0.0206" radius="0.0352"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="0.5" />
        <inertia ixx="0.05" ixy="0" ixz="0" iyy="0.05" iyz="0.0" izz="0.05" />
      </inertial>
    </link>

    <joint name="back_left_wheel_joint" type="continuous">
      <parent link="back_left_motor_link"/>
      <child link="back_left_wheel_link"/>
      <origin xyz="0.0 0.045 -0.015" rpy="1.57 0 0"/>
      <axis xyz="0 0 1"/>
    </joint>
    <link name="back_left_wheel_link">
      <visual>
        <geometry>
          <mesh filename="package://rrt_exploration_tutorials/launch/includes/meshes/wheel.dae"/>
        </geometry>
        <material name="blue">
          <color rgba="0.0 0.0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder length="0.0206" radius="0.0352"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="0.5" />
        <inertia ixx="0.05" ixy="0" ixz="0" iyy="0.05" iyz="0.0" izz="0.05" />
      </inertial>
    </link>

    <joint name="back_right_wheel_joint" type="continuous">
      <parent link="back_right_motor_link"/>
      <child link="back_right_wheel_link"/>
      <origin xyz="0.0 -0.045 -0.015" rpy="1.57 0 0"/>
      <axis xyz="0 0 1"/>
    </joint>
    <link name="back_right_wheel_link">
      <visual>
        <geometry>
          <mesh filename="package://rrt_exploration_tutorials/launch/includes/meshes/wheel.dae"/>
        </geometry>
        <material name="blue">
          <color rgba="0.0 0.0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder length="0.0206" radius="0.0352"/>
        </geometry>
      </collision>
      <inertial>
        <mass value="0.5" />
        <inertia ixx="0.05" ixy="0" ixz="0" iyy="0.05" iyz="0.0" izz="0.05" />
      </inertial>
    </link>
 
    <!-- motor -->
    <joint name="front_left_motor_joint" type="fixed">
      <parent link="base_link"/>
      <child link="front_left_motor_link"/>
      <origin xyz="0.10 0.040 -0.025" rpy="0 0 0"/>
    </joint>
    <link name="front_left_motor_link">
      <visual>
        <geometry>
          <box size="0.04 0.07 0.045" />
        </geometry>
        <material name="blue">
          <color rgba="0.0 0.0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="0.04 0.07 0.045" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.2" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>
    
    <joint name="front_right_motor_joint" type="fixed">
      <parent link="base_link"/>
      <child link="front_right_motor_link"/>
      <origin xyz="0.10 -0.040 -0.025" rpy="0 0 0"/>
    </joint>
    <link name="front_right_motor_link">
      <visual>
        <geometry>
          <box size="0.04 0.07 0.045" />
        </geometry>
        <material name="blue">
          <color rgba="0.0 0.0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="0.04 0.07 0.045" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.2" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>

    <joint name="back_left_motor_joint" type="fixed">
      <parent link="base_link"/>
      <child link="back_left_motor_link"/>
      <origin xyz="-0.10 0.040 -0.025" rpy="0 0 0"/>
    </joint>
    <link name="back_left_motor_link">
      <visual>
        <geometry>
          <box size="0.04 0.07 0.045" />
        </geometry>
        <material name="blue">
          <color rgba="0.0 0.0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="0.04 0.07 0.045" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.2" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>

    <joint name="back_right_motor_joint" type="fixed">
      <parent link="base_link"/>
      <child link="back_right_motor_link"/>
      <origin xyz="-0.10 -0.040 -0.025" rpy="0 0 0"/>
    </joint>
    <link name="back_right_motor_link">
      <visual>
        <geometry>
          <box size="0.04 0.07 0.045" />
        </geometry>
        <material name="blue">
          <color rgba="0.0 0.0 1 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <box size="0.04 0.07 0.045" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.2" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>
    
    <!-- Pillar -->
    
    <joint name="front_left_pillar_joint" type="fixed">
      <parent link="base_link"/>
      <child link="front_left_pillar_link"/>
      <origin xyz="0.11 0.065 0.03" rpy="0 0 0"/>
    </joint>
    <link name="front_left_pillar_link">
      <visual>
        <geometry>
         <cylinder radius="0.005" length="0.06" />
        </geometry>
        <material name="red">
          <color rgba="1 0.0 0.0 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder radius="0.005" length="0.06" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.1" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>

    <joint name="front_right_pillar_joint" type="fixed">
      <parent link="base_link"/>
      <child link="front_right_pillar_link"/>
      <origin xyz="0.11 -0.065 0.03" rpy="0 0 0"/>
    </joint>
    <link name="front_right_pillar_link">
      <visual>
        <geometry>
         <cylinder radius="0.005" length="0.06" />
        </geometry>
        <material name="red">
          <color rgba="1 0.0 0.0 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder radius="0.005" length="0.06" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.1" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>    

    <joint name="back_left_pillar_joint" type="fixed">
      <parent link="base_link"/>
      <child link="back_left_pillar_link"/>
      <origin xyz="-0.11 0.065 0.03" rpy="0 0 0"/>
    </joint>
    <link name="back_left_pillar_link">
      <visual>
        <geometry>
         <cylinder radius="0.005" length="0.06" />
        </geometry>
        <material name="red">
          <color rgba="1 0.0 0.0 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder radius="0.005" length="0.06" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.1" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>

    <joint name="back_right_pillar_joint" type="fixed">
      <parent link="base_link"/>
      <child link="back_right_pillar_link"/>
      <origin xyz="-0.11 -0.065 0.03" rpy="0 0 0"/>
    </joint>
    <link name="back_right_pillar_link">
      <visual>
        <geometry>
         <cylinder radius="0.005" length="0.06" />
        </geometry>
        <material name="red">
          <color rgba="1 0.0 0.0 1" />
        </material>
      </visual>
      <collision>
        <geometry>
          <cylinder radius="0.005" length="0.06" />
        </geometry>
      </collision>
      <inertial>
        <mass value="0.1" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
      </inertial>
    </link>


    <!-- sensors -->
    
    <joint name="laser_joint" type="fixed">
    <parent link="base_second_link" />
    <child link="laser" />
    <origin xyz="0.04 0.0 0.035" rpy="0 0 0"/>
    <axis xyz="0 1 0"/>
</joint>

<link name="laser">
    <visual>
        <geometry>
            <mesh filename="package://rrt_exploration_tutorials/launch/includes/meshes/hokuyo.dae"/> 
        </geometry>
    </visual>
    <inertial>
        <mass value="0.01" />
        <inertia ixx="0.015" ixy="0" ixz="0" iyy="0.0375" iyz="0.0" izz="0.0275" />
    </inertial>
</link>


</robot>
